js/index.js

```diff
@@ -1,10 +1,77 @@
 const isEqual = require("lodash.isequal");
 const expects = require("./expects.js");
 
+function getAllIndexes(arr, val) {
+  var indexes = [], i;
+  for (i = 0; i < arr.length; i++)
+    if (arr[i] === val)
+      indexes.push(i);
+  return indexes;
+}
+
+function checkVertical(puzzle, y, x, word) {
+  for (let i = 0; i < word.length; i++) {
+    if (word[i] !== puzzle[y][x]) {
+      return 0
+    }
+    y++
+  }
+  return 1
+}
+
+function checkDiagonal(puzzle, y, x, word) {
+  let n = 0
+  //check up diagonal
+  let flagup = 1
+  //check down diagonal
+  let flagdown = 1
+  for (let i = 0; i < word.length; i++) {
+    if (y + n >= puzzle.length || x + n >= puzzle[y + n].length || word[i] !== puzzle[y + n][x] || flagdown) {
+      flagdown = 0
+    }
+
+    if (y - n < 0 || x + n >= puzzle.length || word[i] !== puzzle[y - n][x + n]) {
+      flagup = 0
+    }
+    n++;
+  }
+  //to know which end index to compute
+  return flagup ? 1 : (flagdown ? -1 : 0);
+}
 function wordSearch(puzzle, ...words) {
-  // Add your code here
+  const obj = {}
+
+  words.forEach(w => {
+    obj[w] = [];
+    for (let n = 0; n < puzzle.length; n++) {
+      const indices = getAllIndexes(puzzle[n], w[0])
+      indices.forEach(i => {
+        //horizontal
+        if (i + w.length <= puzzle[n].length) {
+          if (puzzle[n].slice(i, w.length + i).join("") === w) {
+            obj[w].push([n, i], [n, i + w.length - 1])
```

Comment:
If you find a match here, you could `break;` the loop to avoid running further calculations on whether the word exists vertically or diagonally. We can assume if we have found the word horizontally we wont need to keep checking. This could potentially save significant time and compute for much larger puzzles or puzzles with many more words. 

You cannot break a `.forEach` loop so easiest next step would be to convert it to a `or (let i of indices)` loop eg:

```
for (let i of indices) {
  //horizontal
  if (i + w.length <= puzzle[n].length) {
    if (puzzle[n].slice(i, w.length + i).join("") === w) {
      obj[w].push([n, i], [n, i + w.length - 1]);
      break;
    }
  }
  //vertical
  if (n + w.length <= puzzle.length) {
      if (checkVertical(puzzle, n, i, w)) {
        obj[w].push([n, i], [n + w.length - 1, i]);
        break;
      }
  }
  //diagonal
  const res = checkDiagonal(puzzle, n, i, w);
  if (res == 1) {
  //down diagonal
    obj[w].push([n, i], [n - w.length + 1, i + w.length - 1]);
    break;
  } else if (res == -1) {
    //down diagonal
    obj[w].push([n, i], [n + w.length - 1, i + w.length - 1]);
    break;
  }
}
```